dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/gist/DefaultGistService.java

```diff
@@ -0,0 +1,163 @@
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.gist;
+
+import static org.hisp.dhis.gist.GistBuilder.createCountBuilder;
+import static org.hisp.dhis.gist.GistBuilder.createFetchBuilder;
+
+import java.util.List;
+import java.util.Map;
+
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+import org.hibernate.Session;
+import org.hibernate.SessionFactory;
+import org.hibernate.query.Query;
+import org.hisp.dhis.schema.RelativePropertyContext;
+import org.hisp.dhis.schema.Schema;
+import org.hisp.dhis.schema.SchemaService;
+import org.hisp.dhis.user.CurrentUserService;
+import org.springframework.stereotype.Service;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+/**
+ * @author Jan Bernitt
+ */
+@Slf4j
+@Service
+@AllArgsConstructor
+public class DefaultGistService implements GistService
+{
+
+    private final SessionFactory sessionFactory;
+
+    private final SchemaService schemaService;
+
+    private final CurrentUserService currentUserService;
+
+    private final ObjectMapper jsonMapper;
+
+    private final GistValidator validator;
+
+    private Session getSession()
+    {
+        return sessionFactory.getCurrentSession();
+    }
+
+    @Override
+    public GistQuery plan( GistQuery query )
+    {
+        return new GistPlanner( query, createPropertyContext( query ) ).plan();
+    }
+
+    @Override
+    public List<?> gist( GistQuery query )
+    {
+        RelativePropertyContext context = createPropertyContext( query );
+        validator.validateQuery( query, context );
+        GistBuilder queryBuilder = createFetchBuilder( query, context, currentUserService.getCurrentUser() );
+        List<Object[]> rows = fetchWithParameters( query, queryBuilder,
+            getSession().createQuery( queryBuilder.buildFetchHQL(), Object[].class ) );
+        queryBuilder.transform( rows );
+        return rows;
+    }
+
+    @Override
+    public GistPager pager( GistQuery query, List<?> rows, Map<String, String[]> params )
+    {
+        int page = 1 + (query.getPageOffset() / query.getPageSize());
+        Schema schema = schemaService.getDynamicSchema( query.getElementType() );
+        String prev = null;
+        String next = null;
+        Integer total = null;
+        if ( schema.haveApiEndpoint() )
+        {
+            String baseURL = GistPager.computeBaseURL( query, params, schemaService::getDynamicSchema );
+            if ( page > 1 )
+            {
+                prev = baseURL + "page=" + (page - 1);
```

Comment:
Perhaps use String.format() to avoid string concatenation. Not important.

Reply:
Uses now URI builder.